src/Razor/src/Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace/RazorWorkspaceListener.cs

```diff
@@ -1,198 +1,35 @@
-            case WorkspaceChangeKind.SolutionRemoved:
-                foreach (var project in e.OldSolution.Projects)
-                {
-                    RemoveProject(project.Id);
-                }
-
-                break;
-
-            default:
-                break;
-        }
+    public RazorWorkspaceListener(ILoggerFactory loggerFactory) : base(loggerFactory.CreateLogger(nameof(RazorWorkspaceListener)))
+    { 
     }
 
-    private void RemoveProject(ProjectId projectId)
+    public void EnsureInitialized(Workspace workspace, string pipeName)
     {
-        ImmutableInterlocked.TryRemove(ref _projectsWithDynamicFile, projectId, out var _);
+        EnsureInitialized(workspace, () => new NamedPipeServerStream(
+                pipeName,
+                PipeDirection.Out,
+                maxNumberOfServerInstances: 1,
+                PipeTransmissionMode.Byte,
+                PipeOptions.CurrentUserOnly | PipeOptions.Asynchronous));
     }
 
-    private void EnqueueUpdate(Project? project)
+    private protected override Task CheckConnectionAsync(Stream stream, CancellationToken cancellationToken)
     {
-        if (_projectInfoFileName is null ||
-            project is not
-            {
-                Language: LanguageNames.CSharp
-            })
+        if (stream is NamedPipeServerStream { IsConnected: false } namedPipe)
         {
-            return;
+            return namedPipe.WaitForConnectionAsync(cancellationToken);
         }
 
-        // Don't queue work for projects that don't have a dynamic file
-        if (!_projectsWithDynamicFile.TryGetValue(project.Id, out var _))
-        {
-            return;
-        }
-
-        var projectId = project.Id;
-        _workQueue.AddWork(projectId);
-    }
-
-    private async ValueTask UpdateCurrentProjectsAsync(ImmutableArray<ProjectId> projectIds, CancellationToken cancellationToken)
-    {
-        var solution = _workspace.AssumeNotNull().CurrentSolution;
-
-        foreach (var projectId in projectIds)
-        {
-            if (_disposeTokenSource.IsCancellationRequested)
-            {
-                return;
-            }
-
-            var project = solution.GetProject(projectId);
-            if (project is null)
-            {
-                _logger?.LogTrace("Project {projectId} is not in workspace", projectId);
-                continue;
-            }
-
-            await SerializeProjectAsync(project, solution, cancellationToken).ConfigureAwait(false);
-        }
-    }
-
-    // private protected for testing
-    private protected virtual Task SerializeProjectAsync(Project project, Solution solution, CancellationToken cancellationToken)
-    {
-        _logger?.LogTrace("Serializing information for {projectId}", project.Id);
-        return RazorProjectInfoSerializer.SerializeAsync(project, _projectInfoFileName.AssumeNotNull(), _logger, cancellationToken);
+        return Task.CompletedTask;
```

Comment:
In what case would we get here?

Reply:
if the pipe already has a connected client